axiom/nr_php_packages.c

```diff
@@ -0,0 +1,142 @@
+/*
+ * Copyright 2022 New Relic Corporation. All rights reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <math.h>
+#include <stdlib.h>
+#include <stdio.h>
+
+#include "nr_axiom.h"
+#include "nr_limits.h"
+#include "nr_php_packages.h"
+#include "util_memory.h"
+#include "util_minmax_heap.h"
+#include "util_vector.h"
+#include "util_hashmap.h"
+#include "util_hashmap_private.h"
+#include "util_strings.h"
+
+nr_php_package_t* nr_php_package_create(char* name, char* version) {
+  nr_php_package_t* p = NULL;
+
+  if (name == NULL || version == NULL) {
+    return NULL;
+  }
+
+  p = (nr_php_package_t*)nr_malloc(sizeof(nr_php_package_t));
+
+  p->package_name = nr_strdup(name);
+  p->package_version = nr_strdup(version);
+
+  return p;
+}
+
+void nr_php_package_destroy(nr_php_package_t* p) {
+  if (p != NULL) {
+    nr_free(p->package_name);
+    nr_free(p->package_version);
+    nr_free(p);
+  }
+}
+
+void nr_php_packages_add_package(nr_php_packages_t** h, nr_php_package_t* p) {
+  if (h == NULL || *h == NULL) {
+    *h = nr_hashmap_create((nr_hashmap_dtor_func_t)nr_php_package_destroy);
```

Comment:
Something is fishy here: if `h` is `NULL`, won't dereferencing with `*h` cause a seg fault?

Reply:
I think C will short circuit if `h == NULL` into the `if` statement body since it is an `||` logical operation.

Reply:
Yes, good catch! I have separated this NULL check into two different lines and check if `h == NULL` before checking `*h == NULL` [24a27c2](https://github.com/newrelic/newrelic-php-agent/pull/754/commits/24a27c2df8aae92a5ab4e6f8693d814d4dc72ff4)